### ISSUE TITLE
An idea to improve Welford implementation in Inductor

### ISSUE THREAD

**Issue by shunting314** [pytorch/pytorch#143337]
### 🐛 Describe the bug

Inductor implements a one-pass Welford algorithm to compute variance. The basic idea is to maintain 3 metrics (weight, mean, m2) for each group of items, and implement a combine function to merge these metrics for two group of items.

- Weight represents the number of elements in the group
- mean represents the mean value of the elements in the group
- M2 represents $\sum(x_i - mean) ^ 2$  summing over each element without the group

There are 3 places this can be improved
1. instead of maintaining mean, we can just maintain sum. By maintaining mean, we need keep adapting the denominator when we combine two groups since the number of elements get changed
2. instead of maintaining M2, we can just maintain sum_of_square $\sum{x_i ^ 2}$ (i.e. equivalent to temporarily pretending the mean is 0). By maintaining M2, we need keep adapting for the fact that 'mean' get changed when we combining two groups.
3. By doing 1 & 2, we don't need track weight anymore. That means we use **LESS REGISTERS**.

The end result is equivalent to leveraging the following equation to compute variance:
$$BiasedVariance = \frac{\sum(x_i - mean) ^ 2}{n} = \frac{\sum{x_i ^ 2}}{n} - mean ^ 2$$

These optimization may not improve perf for every kernel. But I think it worth the effort for the following reasons
1. it simplifies the implementation quite a bit. 
2. if the kernel is fused with surrounding kernels and there is register pressure, using less registers indeed helps perf.

cc @jansel @Chillee @eellison for comments

### Error logs

.

### Versions

.

**Comment by peterbell10** [comment]
Your proposed solution is called the [naive algorithm](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance) and has terrible numerical properties so is not suitable to replace `torch.var`. If you want to use the algorithm, it would need to be an opt-in from the user and you also don't need any special handling for it in inductor since it's just two sums.

Flax's [LayerNorm] for example has a `use_fast_variance` flag which toggles between accurate vs fast variance.

**Comment by Chillee** [comment]
@peterbell10 It's kinda surprising that it's enabled by default 🤔

**Comment by shunting314** [comment]
> you also don't need any special handling for it in inductor since it's just two sums.

Right, the reduction loops of these 2 sums most likely get merged to a single loop


> Your proposed solution is called the [naive algorithm](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance) and has terrible numerical properties so is not suitable to replace torch.var.

The main reason is due to potential overflow of $\sum{x_i ^ 2}$ right? Would it be helpful if do 'divide by N' early enough. E.g. $\sum{\frac{x_i ^ 2}{N}}$ . Then I think there will most likely be no overflow as long as $max_x_i ^ 2$ does not overflow.

**Comment by shunting314** [comment]
I actually don't know how bad the numerical would be if we 
- use fp32 as the accumulators
- divide $xi ^ 2$ by N before the reduction

**Comment by peterbell10** [comment]
It's not just overflow. sum(x**2) - sum(x)**2 is a classic big - big = small precision trap which is prone to catastrophic cancellation. If the input happens to have a mean around 0 you will see reasonable results, but as the mean gets larger compared to the variance you lose more and more precision.

**Comment by shunting314** [comment]
> It's not just overflow. sum(x**2) - sum(x)**2 is a classic big - big = small precision trap which is prone to catastrophic cancellation. If the input happens to have a mean around 0 you will see reasonable results, but as the mean gets larger compared to the variance you lose more and more precision.


I got your general point. But I think there may be still ways to avoid the numerical problem. E.g. 
 `sum(x**2) / N - mean ** 2`  can be computed as `(sqrt(sum(x**2)/N) + mean) * (sqrt(sum(x**2)/N) - mean)` . 

> but as the mean gets larger compared to the variance you lose more and more precision.

I'll find some time to verify the numerical problems on a data set draw from `A_LARGE_NUMBER + torch.rand(shape) * A _SMALL_NUMBER`

**Comment by jansel** [comment]
This seems worth exploring.  Maybe we should add a flag like `config.fast_math` to gate optimization which hurt numerics.
